### ISSUE TITLE
Rule Change: (no-restricted-imports) Resolve to absolute paths.

### ISSUE THREAD

**Issue by infacto** [eslint/eslint#16211]
### What rule do you want to change?

no-restricted-imports

### What change to do you want to make?

Implement suggestions

### Description

It would be great if this rule can treat all paths as absolute paths.
For example paths like `../../bar` > `src/foo/bar`. This makes configuration easier and offers more options especially in a monorepo.

TL;DR: Scroll at the end of this issue. The core issue is that ESLint should resolve relative paths to absolute paths before checking. 

### Example

```
project-root
|- shared
|- projects
   |- server
   |- client
   |- app
```

**projects/server/src/main.ts**

```ts
// The following import paths are NOT allowed.
import { A } from 'projects/client/src/main';
import { B } from 'projects/apps/src/main';
import { C } from '../../client/src/main';
import { D } from '../../apps/src/main';

// The following import paths are allowed.
import { E } from 'shared/main';
import { F } from '../../../shared/main';
import { G } from './utils';
import { H } from 'projects/server/src/utils';
import { I } from 'random-npm-lib';
import { J } from './client/stuff';
import { K } from 'projects/server/src/client/stuff';
```

- Each project is not allowed to import from other projects except its own path and from shared.
- The shared folder is not allowed to import from projects. Only import from its own path.

![monorepo-import](https://user-images.githubusercontent.com/60390085/184863556-ad5a2887-3b5e-4d5c-a092-b1caab89ac42.gif)
(Animated GIF 3s per frame; Green: Allowed import; Red: Not allowed import; Blue: Target)

### Problem

Currently I have not figured out how to achieve this with the currently possible configuration.

I think it could work if I can config absolute paths and this rule resolves all import path to absolute paths.
This means a relative path like `../../client/src/main` is handled as `projects/client/src/main` absolute path.

I have several `.eslintrc.json`. One in root and in each subfolder like shared, server, client and apps. Expecting that each sub eslintrc extends from root. Following configs are pseudo and not tested: 

**projects/server/.eslintrc.json**
```json
{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": ["projects/*", "!project/server"]
      }
    ]
  }
}
```

**projects/client/.eslintrc.json**
```json
{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": ["projects/*", "!project/client"]
      }
    ]
  }
}
```

**projects/app/.eslintrc.json**
```json
{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": ["projects/*", "!project/app"]
      }
    ]
  }
}
```

**shared/.eslintrc.json**
```json
{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "patterns": ["projects/*"]
      }
    ]
  }
}
```

Btw. I use [@typescript-eslint/no-restricted-imports](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-restricted-imports.md) which extends from the ESLint original rule. Just to `allowTypeImports` in TypeScript.

We need an option to resolve all paths to absolute paths.

Basically the core issue is e.g. when a server imports `../../client/*` no error is displayed. Only `project/client/*` is an error. But both paths are absolutly resolves to the `project/client` folder. Therefore this rule should handle all paths as absolute paths. I mean you cannot differ between `../../client` and `../../client` while the first one is absolute `project/client` and the second one is `project/server/client` (imported from another location).

```ts
// This rule should treat both paths as the same path. Because the are the same location (resolved).
import { Something } from '../../client/src/main';
import { Something } from 'projects/client/src/main';
```

What do you think? Can ESLint do that? In NodeJS you would use path resolve.
Alternative is to use tsconfig.json to exclude paths. But afaik it's not possible in an Angular monorepo.

**Comment by mdjermanovic** [comment]
Hi @infacto, thanks for the issue!

The core `no-restricted-imports` rule was designed to work only with bare imports (e.g, imports from `node_modules`). While the configuration for this rule accepts relative paths, `.` and `..` are treated just as directory names.  This rule was never intended to resolve paths and therefore it doesn't really support relative imports. For restricting in-project modules, you should use the `import/no-restricted-paths` plugin rule instead:

https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-restricted-paths.md
